### PR TITLE
Add mp3 support

### PR DESCRIPTION
**Pull request by DanVanAtta**
Change Summary:
- Add a test for ClipPlayer
- Refactor ClipPlayer, simplify it
- Performance fix when muting or unmuting all sounds (3aafc0b)
- Add mp3zoom libraries and update code to be able to play mp3 files

Review note:
- Much of the changes are simply to clean up the code so that it can be updated reasonably
  - bbd55b3b9e97e5294990e4cf92f6711c6290e907 does the actual mp3 support
  - Cleaned up code is mainly public API, or interfaces that have unused methods or excessive and unused parameters
  - The code to test if a sound could be played and to play a sound was copy pasted, one clean up was to combine this functionality so only one code path actually then plays sounds

To test the performance fix:
- start a game, go to sound options
- use the select all or select none check boxes, click okay. Notice that the UI will freeze up for a few seconds (FYI, with a fast HD you may not notice this delay - after upgrading to a SSD disk that is a full 1000x faster than the Sata drive it replaced, I no longer noticed this delay)

To test mp3 file support:
- Make sure you can hear game sounds first (Napoleonic start sound is my favorite)
- Check out this branch:
  - `git remote add DanV https://github.com/DanVanAtta/triplea.git`
  - `git fetch DanV`
  - `git checkout clipPlayer`
- Add the three new jar libraries to your eclipse path (mp3*.jar, jl1*.jar, tritonus*.jar)
- Repeat the first experiment, this confirms wav files still play
- Use your favorite mp3 encoding utility to convert the test sound to mp3. `lame` is the tool I used, the napolean start sound is under "assets/preindustrial/game_start". Audacity is another tool I think available for windows that could do the same thing with a GUI (it comes with a lame plugin)

Two gotchyas for mp3 file testing:
- Adding the jar libraries is needed at runtime, not compile time. So you won't get any compile errors if they are not on your path or any other kind of reminder to do that.
- If sound is muted in the game then you won't hear anything, both the sound option needs to be unmuted and "sounds enabled" checked! A few times after coming back from playing I ran into that the hard way.

[Review on Reviewable](https://reviewable.io/reviews/triplea-game/triplea/201)

### CODE REVIEW

File: src/games/strategy/sound/ClipCache.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package games.strategy.sound;
```
Comment (DanVanAtta) [inline src/games/strategy/sound/ClipCache.java]: ClipCache was extracted from ClipPlayer, then simplified slightly.

**Comment by ron-murhammer** [comment]
@DanVanAtta Looks like this needs rebased.

**Comment by DanVanAtta** [comment]
Rebased

**Comment by DanVanAtta** [comment]
This is actually not that difficult of a change to review. A person could cover most of it in 10 minutes. Except for the second to last change all that is happening is we are cleaning up the sound playing code, which was in a pretty messy state. In fact, this change does nothing functionally, but it does allow us to drop in mp3 files and they should get played similar to wav or the other support formats.

**Comment by DanVanAtta** [comment]
Comments on how this can be reviewed and or merged? Spending about minutes on the overall changes and then 15-30 minutes on ea10cc4 would be my suggestion for reviewing this PR. 

Regardless, it would be helpful if you could share your expectations and intent as far as this PR is concerned @veqryn  and @ron-murhammer .

**Comment by ron-murhammer** [comment]
@DanVanAtta Overall, looks pretty good and most of the comments are minor. There are a few though that need addressed before this can be merged.

**Comment by DanVanAtta** [comment]
@ron-murhammer , thanks for the review, I think all of the "must-fix" stuff should now be addressed.

**Comment by ron-murhammer** [comment]
@DanVanAtta The only remaining question I have is about this line:

AudioFormat decodedFormat = new AudioFormat(AudioFormat.Encoding.PCM_SIGNED, format.getSampleRate(), 16, format.getChannels(), format.getChannels() * 2, format.getSampleRate(), false);
 
Can you at least describe why these parameters are set to the values in this thread? Just want to understand what's going on here.

**Comment by DanVanAtta** [comment]
"Can you at least describe why these parameters are set to the values in this thread? Just want to understand what's going on here."

```
AudioFormat decodedFormat = new 
AudioFormat(AudioFormat.Encoding.PCM_SIGNED, format.getSampleRate(), 16, format.getChannels(), format.getChannels() * 2, format.getSampleRate(), false);
```

Now fixed in:  3f38ceb and 849e2ab

**Comment by DanVanAtta** [comment]
Except for messing up the commit history a few times and merge conflict resolution, review points I think have all been addressed now (@ron-murhammer)

**Comment by ron-murhammer** [comment]
@DanVanAtta - Alright refactors look good. Merging this now.

**Comment by veqryn** [comment]
Can the added `jars` use the existing jar format please?

File: src/games/strategy/engine/framework/AbstractGameLoader.java [hunk 2]
```diff
@@ -15,10 +15,8 @@
 
   @Override
   public void shutDown() {
-    if (m_soundChannel != null) {
+    if (m_game != null) {
```
Comment (veqryn) [inline src/games/strategy/engine/framework/AbstractGameLoader.java]: This change will cause errors if soundChannel is null

File: src/games/strategy/sound/ClipPlayer.java [hunk 3]
```diff
@@ -442,239 +419,67 @@ private void parseClipPaths(final String pathName, final boolean subFolder) {
...
     }
-    if (thisSoundFile == null || !thisSoundFile.exists()) {
-      // final long startTime = System.currentTimeMillis();
-      // we are probably using zipped sounds. there might be a better way to do this...
-      final String soundFilePath = thisSoundURL.getPath();
-      if (soundFilePath != null && soundFilePath.length() > 5 && soundFilePath.indexOf(".zip!") != -1) {
-        // so the URL with a zip or jar in it, will start with "file:", and unfortunately when you make a file and test
-        // if it exists, if it
-        // starts with that it doesn't exist
-        final int index1 = Math.max(0, Math.min(soundFilePath.length(),
```
Comment (veqryn) [inline src/games/strategy/sound/ClipPlayer.java]: It looks like you are removing all logic needed to get a sound clip out of a map zip??

**Comment by veqryn** [comment]
I think this was merged prematurely.
There are no comments or explanation about the giant chunk of code being removed that deals with sounds that are inside of a map's zip.

@DanVanAtta can you explain what is going on, and why that code was removed, and whether this whole thing still works with zips?

File: src/games/strategy/engine/framework/AbstractGameLoader.java (hunk 2 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/AbstractGameLoader.java]: Fix is in PR #314

**Comment by DanVanAtta** [comment]
"Can the added jars use the existing jar format please?"

Could you please clarify? The added jars are interestingly enough 'run-time' dependencies and not 'compile-time' dependencies. Hence they only needed to be added to the lib folder and nothing further.

**Comment by veqryn** [comment]
For the jar format, I only meant `<name>-<version>.jar`
It looks like the new jars differ from this slightly, and one of them is missing a version.

**Comment by veqryn** [comment]
@DanVanAtta: so, why is all the code to parse sounds out of a map zip being removed?

**Comment by DanVanAtta** [comment]
I do not think there is a common understanding of what was removed with respect to that large code chunk. Sound files included with a map zip, located under the "sounds" folder, will still play.  For example, with a star wars galactic war zip, this sound file does still play:  ./star_wars_galactic_war/map/sounds/game_start.wav.

What was removed is the ability to bundle *only* the audio files together in a zip. This is not used for any of the maps that are included with the game, nor is it used for any of the extra zipped maps. Not only is the feature unused, but it is unnecessary as well, it implies the abilitiy to have a "zip-in-a-zip" for sound files. Sound files are better left unzipped rather than zipped generally, there is memory and processing overhead for the compression, and almost no space is saved when zipping them. Thus, a "zip-in-a-zip" for sound files really does not make sense, thus the feature is unnecessary.

**Comment by DanVanAtta** [comment]
@veqryn , please let me know if that explanation does not clarify things. 

I opened #317 to discuss and track the jar naming issue.

**Comment by veqryn** [comment]
I don't believe we ever allowed zip within zip functionality.  That code, which I and @frigoref wrote, was purely to allow accessing sounds within a zipped map file.
If it is working right now without that code, then some magic must have happened between the time we wrote it and now...

File: src/games/strategy/sound/ClipPlayer.java (hunk 3 above)
Comment (DanVanAtta) [inline src/games/strategy/sound/ClipPlayer.java]: All of this code is never executed. Debugging the what remains, it looks like the only active code in this whole method is this:
```
      if (testClipSuccessful(thisSoundURL)) {
        availableSounds.add(thisSoundURL);
      }
```
For the wav files in a map zip case, the parameter value will look something like this:
```
jar:file:/home/dan/triplea/maps/star_wars_galactic_war.zip!/sounds/click2.wav
```
The code in CiipPlayer then has no problem creating a Clip object with that.

I would surmise then that it was never realized that this code was all dead code effectively, that the fall back logic executed later was the only logic that was really doing anything.

For reference, debugging this, a stack trace to get to this point looks like:
```
ClipPlayer.createAndAddClips(String) line: 403	
ClipPlayer.parseClipPaths(String, boolean) line: 378	
ClipPlayer.loadClipPath(String, boolean, boolean) line: 333	
ClipPlayer.loadClip(String, String, boolean) line: 322	
ClipPlayer.preLoadClip(String) line: 303	
SoundPath.preLoadSounds(SoundPath$SoundType) line: 144	
ClipPlayer.getInstance(ResourceLoader, GameData) line: 140	
UIContext$1.run() line: 98	
Thread.run() line: 745
```

Stack trace with method signatures and a bit more context:
```
      SoundPath.preLoadSounds(SoundPath.SoundType.GENERAL);
        clipPlayer.preLoadClip(CLIP_CLICK_PLOT);

  protected void preLoadClip(final String clipName) {
    loadClip(clipName, null, true);

  private synchronized Clip loadClip(final String clipName, final String subFolder, final boolean parseThenTestOnly) {
        return loadClipPath(clipName, false, parseThenTestOnly);

  private Clip loadClipPath(final String pathName, final boolean subFolder, final boolean parseThenTestOnly) {
    if (!sounds.containsKey(pathName)) {
      // parse sounds for the first time
      parseClipPaths(pathName, subFolder);

  private void parseClipPaths(final String pathName, final boolean subFolder) {
      :
      availableSounds.addAll(createAndAddClips(ASSETS_SOUNDS_FOLDER + "/" + path));


  protected List<URL> createAndAddClips(final String resourceAndPathURL) {
   
```

And then eventually it looks like most of the time the code short circuits searching for possible paths, and those that do exist, then eventually hit the "testClipSuccessful" logic which is able to load the sound file, even when inside of a compressed file.

**Comment by DanVanAtta** [comment]
@veqryn , please see the in-line comment about the sound zip code, I included details from a debug session.

Otherwise is there anything else you would have like done before this was merged? My next question, in your opinion, where exactly is the point of diminishing returns when it comes to reviewing a PR?

Last, as discussion, I strongly feel that we must have a relatively agile process:
- It's nearly impossible to not have any bugs in this code base. If we try to go for 100%, we will never get anything done. I'm not saying let's shoot from the hip and not care, but achieving 97.5%+ is okay. 
- On some of these PRs, we have spent magnitudes more time reviewing than it actually took to do everything else up to that point. If work scales like that, then what happens when a month or two worth of work is up for review? Is it really reasonable to have a review process that needs a couple *years* to review large changes? The answer to this one I don't think can just be "don't submit large changes", but rather we need to be sure our review process takes much less time than it does to actually code the feature in the first place. I think we do lack mechanisms and perhaps process to help ensure code review is fast, efficient, and effective relative to risks and cost.

Finally, the reason I feel most strongly about moving somewhat quickly is because the tripleA community is clearly shrinking. It's getting very difficult to start lobby games. Thus, I'm afraid if we can't make enough progress fast enough, there won't be any community left playing this game to have made the code updates worth it. So I believe we need to strike a balance between moving quickly, and moving safely.

**Comment by DanVanAtta** [comment]
#319 opened to talk further about process improvements to review and merge code more quickly.

**Comment by veqryn** [comment]
I have no issues with moving quickly, but a giant chunk of code was deleted without even a comment or hint as to why, or a question from anyone else as to why.  If that had of been specified a little better, then I don't think I would have even commented on this PR at all.

**Comment by DanVanAtta** [comment]
> "if that had of been specified a little better, then I don't think I would have even commented on this PR at all."

I think part of the problem is the lack of commentary and difficulty having incremental/partial reviews of the larger PRs.

I would wonder if there is a way we could have had those questions about the code changes asked sooner. It's very reasonable to ask why seemingly necessary code is being removed and there is no guarantee that I overlooked something (as I had twice pointed out by @ron-murhammer ).

**Comment by DanVanAtta** [comment]
@veqryn , for the removed code, a better, more focused commit with message would have been sufficient? Ultimately I also wonder if this should have been many PRs based on each other.

**Comment by veqryn** [comment]
Honestly I am not too sure.  I am still learning and working through this too...

**Pull request merged**